### PR TITLE
 Two windings transformer: add phaseAngleClock, only one tap changer regulating.

### PR DESCRIPTION
**Pull request by marqueslanauja**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
More than one tap changer regulating control enabled is allowed (ratio and phase)

**What is the new behavior (if this is a feature change)?**
Only one tap changer regulating control enabled is allowed
`PhaseAngleClock` is added in `twoWindingsTransformer` as an extension

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)
Some tests has been adjusted as they have `twoWindingsTransformers` with more than one `tapChanger` control enabled.

### CODE REVIEW

**Comment by zamarrenolm** [comment]
sonar warning about duplicated code is related to common methods with other XML serializers. It is not relevant. 

One option to fix the duplicated code would be to set the name of the XSD resource, the namespace URI and the namespace prefix as attributes at `ExtensionXmlSerializer`, so methods to obtain these items can be implemented at `ExensionXmlSerializer`, without the need to override at every child class.

**Comment by MioRtia** [comment]
> sonar warning about duplicated code is related to common methods with other XML serializers. It is not relevant.
> 
> One option to fix the duplicated code would be to set the name of the XSD resource, the namespace URI and the namespace prefix as attributes at `ExtensionXmlSerializer`, so methods to obtain these items can be implemented at `ExensionXmlSerializer`, without the need to override at every child class.

For now, we consider this as false positive, it will not block the merging into master for us.

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/TwoWindingsTransformerConversion.java [hunk 1]
```diff
@@ -77,6 +80,20 @@ public void convert() {
...
         addTapChangers(tx);
+
+        int phaseAngleClock1 = end1.asInt(STRING_PHASE_ANGLE_CLOCK, 0);
+        int phaseAngleClock2 = end2.asInt(STRING_PHASE_ANGLE_CLOCK, 0);
+
+        // add phaseAngleClock as an extension, cgmes does not allow pac at end1
+        if (phaseAngleClock1 != 0) {
+            String reason = "Unsupported modelling: two winding transformer with phase angle clock at end1";
+            invalid(reason);
+            throw new PowsyblException(String.format("TwoWindingTransformer %s %s", tx.getId(), reason));
```
[the remark this answers is not in the document]
Comment (marqueslanauja) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/TwoWindingsTransformerConversion.java]: You are right. Done.

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/PhaseAngleClockTwoWindingsTransformer.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+package com.powsybl.iidm.network.extensions;
+
+import com.powsybl.commons.PowsyblException;
+import com.powsybl.commons.extensions.AbstractExtension;
+import com.powsybl.iidm.network.TwoWindingsTransformer;
+
+/**
+ * @author José Antonio Marqués <marquesja at aia.es>
+ */
+public class PhaseAngleClockTwoWindingsTransformer extends AbstractExtension<TwoWindingsTransformer> {
```
[the remark this answers is not in the document]
Comment (marqueslanauja) [reply iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/PhaseAngleClockTwoWindingsTransformer.java]: Done

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/PhaseAngleClockTwoWindingsTransformerXmlSerializer.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.io.InputStream;
+
+import javax.xml.stream.XMLStreamException;
+
+/**
+ * @author José Antonio Marqués <marquesja at aia.es>
+ */
+@AutoService(ExtensionXmlSerializer.class)
+public class PhaseAngleClockTwoWindingsTransformerXmlSerializer
```
[the remark this answers is not in the document]
Comment (marqueslanauja) [reply iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/PhaseAngleClockTwoWindingsTransformerXmlSerializer.java]: Done

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/PhaseAngleClockTwoWindingsTransformerXmlSerializer.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Override
+    public String getNamespaceUri() {
+        return "http://www.powsybl.org/schema/iidm/ext/phase_angle_clock_two_windings_transformer/1_0";
+    }
+
+    @Override
+    public String getNamespacePrefix() {
+        return "pac2wt";
```
[the remark this answers is not in the document]
Comment (marqueslanauja) [reply iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/PhaseAngleClockTwoWindingsTransformerXmlSerializer.java]: Prefix cannot begin with a number. I have used `twowtpac`. Done

File: iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/PhaseAngleClockTwoWindingsTransformerXmlSerializer.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Override
+    public PhaseAngleClockTwoWindingsTransformer read(TwoWindingsTransformer extendable, XmlReaderContext context) throws XMLStreamException {
+        int phaseAngleClock = XmlUtil.readOptionalIntegerAttribute(context.getReader(), "phaseAngleClock", 0);
+        return new PhaseAngleClockTwoWindingsTransformer(extendable, phaseAngleClock);
+    }
+
+    @Override
+    public String getExtensionName() {
+        return "phaseAngleClockTwoWindingsTransformer";
```
[the remark this answers is not in the document]
Comment (marqueslanauja) [reply iidm/iidm-extensions/src/main/java/com/powsybl/iidm/network/extensions/PhaseAngleClockTwoWindingsTransformerXmlSerializer.java]: Done

File: iidm/iidm-extensions/src/main/resources/xsd/phaseAngleClockTwoWindingsTransformer.xsd [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+-->
+<xs:schema version="1.0"
+           xmlns:xs="http://www.w3.org/2001/XMLSchema"
+           targetNamespace="http://www.powsybl.org/schema/iidm/ext/phase_angle_clock_two_windings_transformer/1_0"
+           elementFormDefault="qualified">
+    <xs:element name="phaseAngleClockTwoWindingsTransformer">
```
[the remark this answers is not in the document]
Comment (marqueslanauja) [reply iidm/iidm-extensions/src/main/resources/xsd/phaseAngleClockTwoWindingsTransformer.xsd]: Done

File: iidm/iidm-extensions/src/main/resources/xsd/phaseAngleClockTwoWindingsTransformer.xsd [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        <xs:complexType>
+            <xs:attribute name="phaseAngleClock" use="optional" type="xs:integer"/>
```
[the remark this answers is not in the document]
Comment (marqueslanauja) [reply iidm/iidm-extensions/src/main/resources/xsd/phaseAngleClockTwoWindingsTransformer.xsd]: You are right. Done.

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/ValidationUtil.java [hunk 8]
```diff
@@ -316,6 +319,16 @@ static void checkPhaseTapChangerRegulation(Validable validable, PhaseTapChanger.
         }
     }
 
+    public static void checkOnlyOneTapChangerRegulatingEnabled(Validable validable,
+        Set<TapChanger> tapChangersNotIncludingTheModified, boolean regulating) {
+        if (regulating) {
+            long enabled = tapChangersNotIncludingTheModified.stream().filter(o -> o.isRegulating()).count();
```
[the remark this answers is not in the document]
Comment (marqueslanauja) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/ValidationUtil.java]: Done

**Comment by geofjamg** [comment]
Should it be flagged as breaking change (and added to migration guide) as for instance load flow impl have to be modified to take this attribute into account?

**Comment by MioRtia** [comment]
> Should it be flagged as breaking change (and added to migration guide) as for instance load flow impl have to be modified to take this attribute into account?

I am not sure this PR qualifies as Breaking Change since no breaking change is explicitly made here... Behavior is not modified anywhere by this PR. Maybe in another PR, a LF parameter indicating if phase angle clocks are ignored or not will be added and this PR should then be declared as broken change but not this one in my opinion.

**Comment by zamarrenolm** [comment]
> > Should it be flagged as breaking change (and added to migration guide) as for instance load flow impl have to be modified to take this attribute into account?
> 
> I am not sure this PR qualifies as Breaking Change since no breaking change is explicitly made here... Behavior is not modified anywhere by this PR. Maybe in another PR, a LF parameter indicating if phase angle clocks are ignored or not will be added and this PR should then be declared as broken change but not this one in my opinion.

I agree with Miora. Phase angle clock values are relevant for short-circuit analysis and load flow calculations can safely ignore them (voltages and flows obtained will be the same with or without these values). 

Some TSOs take them into account in their load flow calculations, but it is not a requirement. In these cases load flow results angles have a phase shift in some buses (after the transformer end that has the phase angle clock). As these buses are not connected to other parts of the network, this phase shift has no real impact on the electric analysis using load flow. Storing phase angle clocks has been introduced mainly to be able to do comparisons in these situations.

Next pull request related to this set of changes will be a breaking change, as it will introduce the new three winding transformer API.

**Comment by geofjamg** [comment]
> > > Should it be flagged as breaking change (and added to migration guide) as for instance load flow impl have to be modified to take this attribute into account?
> > 
> > 
> > I am not sure this PR qualifies as Breaking Change since no breaking change is explicitly made here... Behavior is not modified anywhere by this PR. Maybe in another PR, a LF parameter indicating if phase angle clocks are ignored or not will be added and this PR should then be declared as broken change but not this one in my opinion.
> 
> I agree with Miora. Phase angle clock values are relevant for short-circuit analysis and load flow calculations can safely ignore them (voltages and flows obtained will be the same with or without these values).
> 
> Some TSOs take them into account in their load flow calculations, but it is not a requirement. In these cases load flow results angles have a phase shift in some buses (after the transformer end that has the phase angle clock). As these buses are not connected to other parts of the network, this phase shift has no real impact on the electric analysis using load flow. Storing phase angle clocks has been introduced mainly to be able to do comparisons in these situations.
> 
> Next pull request related to this set of changes will be a breaking change, as it will introduce the new three winding transformer API.

Ok thanks.

**Pull request merged**